### DISCUSSION TITLE
SSHPASS

### DISCUSSION THREAD

**Discussion by brsylmz23** [Xilinx/finn#389] [Q&A]
Hi, 

I am trying to complete cnv_end2end example and I could not understand how to fill this line about sshpass. 
![Ekran Alıntısı](https://user-images.githubusercontent.com/41904237/136997708-789b7c4a-a490-4783-9f0a-4674610d6763.PNG)

**Comment by heborras** [comment]
Hi @brsylmz23 ,
this line is supposed to establish a connection to a locally connected PYNQ-Z1 board and to print out the contents of the target folder, where the deployment files should have been transferred to. If you have a PYNQ-Z1 connected via the network to your computer you can tell FINN what the access details of the PYNQ-Z1 are and where to find it by setting the following environment variables:

- `PYNQ_IP`
- `PYNQ_USERNAME`
- `PYNQ_PASSWORD`
- `PYNQ_PORT`

The documentation gives a short description what each of these environment variables mean: https://finn.readthedocs.io/en/latest/getting_started.html#environment-variables

**Reply by brsylmz23** [reply]
Sorry, I thought I fix this issue and marked as answer and now I have some questions too. 
When using ssh , we need to fill this line as {username}@{ip}. So this "ip" means PYNQ-Z2's ip right? Or do I need to use different ip?

**Reply by rpitonak** [reply]
Hi @brsylmz23,

Yes, it is the IP address of the target board (PYNQ-Z2 in your case I suppose). You can see that in the example notebook the value is read from the ENV variable `PYNQ_IP` and if the variable is not set, [the default value of PYNQ boards](https://pynq.readthedocs.io/en/latest/getting_started/pynq_z2_setup.html#network-connection) is used.
```
ip = os.getenv("PYNQ_IP", "192.168.2.99")
```

**Reply by brsylmz23** [reply]
Thanks for the answer. When I try to use my board's IP address. I see this error. I tried some solutions on blogs but I could not fix it. 
![image](https://user-images.githubusercontent.com/41904237/138740871-f81f68ac-8244-4177-9f91-0a3981d4a153.png)

**Reply by rpitonak** [reply]
Be sure that you either use `-i` option of ssh command with the path to SSH keys, or that keys are in `~/.ssh` directory in the container. Also follow this tutorial: https://finn.readthedocs.io/en/latest/getting_started.html#pynq-board-first-time-setup to generate a new pair of ssh keys, if you haven't done yet.

**Reply by brsylmz23** [reply]
Actually I can reach PYNQ from terminal but cant reach on notebook.
![image](https://user-images.githubusercontent.com/41904237/138877273-be35cc6d-0d7e-4fb4-babb-0c2aa5e155d2.png)
![image](https://user-images.githubusercontent.com/41904237/138877408-ddb46196-282e-4044-8afe-db8edf1b998f.png)

**Reply by rpitonak** [reply]
This error message means that you are connected to the board, just the deployment repository does not exist there. Try to remove the `{target_dir_pynq}` to verify that you can execute plain `ls` on board.

**Reply by brsylmz23** [reply]
Thanks for your help, you are right. But still there is some failure. Is it okay without {target_dir_pynq} or how can I configure repository correctly?
![image](https://user-images.githubusercontent.com/41904237/138937229-8735b790-cc2d-4c0a-a24c-d9f9eeeef939.png)

**Reply by rpitonak** [reply]
I am sorry, I can't help you if I do not see the output of previous cells. Can you provide more information about what have you done so far?

Also, where exactly is `sshpass` command used? I can't see it in https://github.com/Xilinx/finn/blob/master/notebooks/end2end_example/bnn-pynq/cnv_end2end_example.ipynb which you have mentioned above.

**Reply by brsylmz23** [reply]
Sure. I guess I was using old cnv_end2end_example, so I modified my codes according to what you sent. I created a key by following these steps(https://finn-dev.readthedocs.io/en/latest/getting_started.html#pynq-board-first-time-setup). After that I tested and I can access ssh PYNQ card from my terminal as I shared above. And now after modifying my codes again, I got this error. Can u tell me more specific what you need to see to get the problem. Thanks.
![image](https://user-images.githubusercontent.com/41904237/138957043-e673e60b-4d38-4459-bd15-1a411e994331.png)

**Reply by rpitonak** [reply]
If you followed the tutorial and you generated the keys please try one of these two options:
a) Update `options` variable so it looks like this:
```
options = " -i <PATH_TO_GENERATED_KEYS>/id_rsa -o PreferredAuthentications=publickey -o PasswordAuthentication=no"
```
> \<PATH_TO_GENERATED_KEYS\> shall be replaced with absolute path to directory with the keys.

b) Execute to the docker container and try to copy the keys to `~/.ssh/`( in your case `/home/baris/.ssh/`)

Please be aware that both of this steps should be done inside of the docker container, that means that `PATH_TO_GENERATED_KEYS` must be absolute path in the docker container or `~/.ssh/` must contain the keys inside of the docker container.

**Reply by brsylmz23** [reply]
a) I updated "options" according to directory of my key which is inside Docker and under /.ssh.
![image](https://user-images.githubusercontent.com/41904237/139058017-c9ba3d7f-b249-4d7a-a40c-8676cd72a2cf.png)
Also same directory on notebook but still gives some permission error.
![image](https://user-images.githubusercontent.com/41904237/139068957-4aad1060-e2de-48a1-8241-6e7a3ddaf3b4.png)

And these are my permissions.
![image](https://user-images.githubusercontent.com/41904237/139069147-d7802472-34cf-40d6-957e-8dd94fb39097.png)

**Reply by rpitonak** [reply]
They should rather be placed in`/home/baris/.ssh` because that is the home directory of user set in docker container

**Reply by brsylmz23** [reply]
Thanks, It worked like u said. But I have still problem on this line. My "pynq_deployment_**" folder always created on my tmp folder, not in '/home/xilinx/finn_dev_baris'
![image](https://user-images.githubusercontent.com/41904237/139077954-8173ade6-75c2-45f2-868a-65e143e1ceee.png)
![image](https://user-images.githubusercontent.com/41904237/139078071-9e304f60-1521-426b-a05d-a30e4e658b39.png)

**Reply by rpitonak** [reply]
Can you post the output of these 3 cells?
<img width="1205" alt="Screenshot 2021-10-27 at 16 06 59" src="https://user-images.githubusercontent.com/26160778/139082145-c5081c9f-bed5-4d4d-8330-6de85a96e889.png">

**Reply by brsylmz23** [reply]
Sure.
![image](https://user-images.githubusercontent.com/41904237/139082773-4c39e22c-c0d2-42ae-8f50-d18efd44b9ba.png)

**Reply by rpitonak** [reply]
Weird, it should be copied with this command:
```
model = model.transform(DeployToPYNQ(ip, port, username, password, target_dir))
```

to target_dir on board.
